iree/compiler/Conversion/LinalgToSPIRV/FlattenMemRefSubspanPass.cpp

```diff
@@ -0,0 +1,329 @@
+                                                linearIndex);
+    return success();
+  }
+};
+
+/// Linearizes indices in memref.store ops.
+struct LinearizeStoreIndices final
+    : public OpConversionPattern<memref::StoreOp> {
+  using OpConversionPattern<memref::StoreOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      memref::StoreOp storeOp, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    typename memref::StoreOp::Adaptor adaptor(operands);
+    assert(isRankZeroOrOneMemRef(adaptor.memref().getType()));
+
+    Value linearIndex =
+        linearizeIndices(storeOp.getMemRefType(), storeOp.getIndices(),
+                         storeOp.getLoc(), rewriter);
+    rewriter.replaceOpWithNewOp<memref::StoreOp>(storeOp, adaptor.value(),
+                                                 adaptor.memref(), linearIndex);
+    return success();
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Folding Patterns
+//===----------------------------------------------------------------------===//
+
+/// Returns the number of bytes of the given `type`. Returns llvm::None if
+/// cannot deduce.
+///
+/// Note that this should be kept consistent with how the byte offset was
+/// calculated in the subspan ops!
+Optional<int64_t> getNumBytes(Type type) {
+  if (type.isIntOrFloat()) return (type.getIntOrFloatBitWidth() + 7) / 8;
+  if (auto vectorType = type.dyn_cast<VectorType>()) {
+    auto elementBytes = getNumBytes(vectorType.getElementType());
+    if (!elementBytes) return llvm::None;
+    return elementBytes.getValue() * vectorType.getNumElements();
+  }
+  return llvm::None;
+}
+
+/// Folds the byte offset on subspan ops into the consumer load/store ops.
+template <typename OpType>
+struct FoldSubspanOffsetIntoLoadStore final : public OpRewritePattern<OpType> {
+  using OpRewritePattern<OpType>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(OpType op,
+                                PatternRewriter &rewriter) const override {
+    auto memrefType = op.memref().getType().template cast<MemRefType>();
+    if (!isRankZeroOrOneMemRef(memrefType)) return failure();
+
+    auto subspanOp =
+        op.memref()
+            .template getDefiningOp<IREE::HAL::InterfaceBindingSubspanOp>();
+    if (!subspanOp) return failure();
+
+    // If the subspan op has a zero byte offset then we are done.
+    if (matchPattern(subspanOp.byte_offset(), m_Zero())) return failure();
+    // byte length is unsupported for now.
+    if (subspanOp.byte_length()) return failure();
+
+    // Create a new subspan op with zero byte offset.
+    Value zero = rewriter.create<ConstantIndexOp>(op.memref().getLoc(), 0);
+    Value newSubspan = rewriter.create<IREE::HAL::InterfaceBindingSubspanOp>(
+        op.memref().getLoc(), subspanOp.getType(), subspanOp.binding(), zero,
+        subspanOp.byte_length());
```

Comment:
[optional] I recently learned that there is `ImplicitLocOpBuilder b(op.getLoc(), rewriter);`. With the builder, you don't have to pass a `Location` to the builder anymore.